### PR TITLE
adjustments to the latest inlining interface changes

### PR DESCRIPTION
**Pull request by aviatesk**
Still needs to adjust to https://github.com/JuliaLang/julia/pull/52415.

### CODE REVIEW

**Comment by wsmoses** [comment]
@vchuravy is this now subsumed by recent 1.11 stuff?

**Comment by vchuravy** [comment]
No I haven't touched any of that

**Comment by wsmoses** [comment]
@vchuravy can you take a look at this. I think this is the next pending thing for 1.11 support (with the incoming jll bump: 
https://github.com/JuliaPackaging/Yggdrasil/pull/9354

**Comment by aviatesk** [comment]
I tried updating it, but it seems we might need a fix in Base. Specifically, because of JuliaLang/julia/#52064, the `inlining_policy` in Base no longer requires `mi`, so `mi` isn't being passed to `inlining_policy` anymore. However, Enzyme's inlining rule still needs it. Since we can't accurately recover `mi` from `info` or `src`, a change to Base might be necessary.

**Comment by aviatesk** [comment]
When does this error usually occur?
```
Error During Test at /Users/aviatesk/julia/packages/Enzyme/test/abi.jl:4
  Got exception outside of a @test
  
  No augmented forward pass found for julia.gc_loaded
   at context:   %37 = call {} addrspace(10)* addrspace(13)* @julia.gc_loaded({} addrspace(10)* noundef %17, {} addrspace(10)** noundef %36) #12, !dbg !39
  
  Stacktrace:
   [1] getindex
     @ ./essentials.jl:916
   [2] squareRetArray
     @ ~/julia/packages/Enzyme/test/abi.jl:86
  
  
  Stacktrace:
    [1] getindex
      @ ./essentials.jl:916 [inlined]
    [2] squareRetArray
      @ ~/julia/packages/Enzyme/test/abi.jl:86 [inlined]
    [3] diffejulia_squareRetArray_29137wrap
      @ ~/julia/packages/Enzyme/test/abi.jl:0
    [4] macro expansion
      @ ~/julia/packages/Enzyme/src/compiler.jl:7179 [inlined]
    [5] enzyme_call
      @ ~/julia/packages/Enzyme/src/compiler.jl:6788 [inlined]
    [6] CombinedAdjointThunk
      @ ~/julia/packages/Enzyme/src/compiler.jl:6665 [inlined]
    [7] autodiff
      @ ~/julia/packages/Enzyme/src/Enzyme.jl:320 [inlined]
    [8] autodiff(mode::ReverseMode{false, FFIABI, false, false}, f::var"#squareRetArray#28", ::Type{Const}, args::Duplicated{Vector{Float64}})
      @ Enzyme ~/julia/packages/Enzyme/src/Enzyme.jl:332
    [9] macro expansion
      @ ~/julia/packages/Enzyme/test/abi.jl:91 [inlined]
   [10] macro expansion
      @ ~/julia/julia5/usr/share/julia/stdlib/v1.11/Test/src/Test.jl:1700 [inlined]
   [11] top-level scope
      @ ~/julia/packages/Enzyme/test/abi.jl:6
   [12] include(fname::String)
      @ Main ./sysimg.jl:38
   [13] top-level scope
      @ ~/julia/packages/Enzyme/test/runtests.jl:85
   [14] include(fname::String)
      @ Main ./sysimg.jl:38
   [15] top-level scope
      @ none:6
   [16] eval
      @ ./boot.jl:430 [inlined]
   [17] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:291
   [18] _start()
      @ Base ./client.jl:526
```

**Comment by wsmoses** [comment]
That is fixed by the jll bump -- which presently seems trapped in yggdrassil hell: https://github.com/JuliaPackaging/Yggdrasil/pull/9354

cc @vchuravy @staticfloat @giordano

**Comment by wsmoses** [comment]
@aviatesk okay a jll should now be available

**Comment by aviatesk** [comment]
I have switched to an approach using the new `CallInfo`. This now should work without any changes on the Base side. However, I have noticed the following error occurring in the assertion build (though I am not sure if it is caused by this PR):
```
Test Summary:    | Pass  Total  Time
Persistent tasks |    1      1  9.0s
Assertion failed: ((IsIntAttr || Attribute::isEnumAttrKind(Kind)) && "Not an enum or int attribute"), function get, file /workspace/srcdir/llvm-project/llvm/lib/IR/Attributes.cpp, line 95.

[18495] signal 6: Abort trap: 6
in expression starting at /Users/aviatesk/julia/packages/Enzyme/test/abi.jl:4
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Allocations: 151710873 (Pool: 151707636; Big: 3237); GC: 101
```

File: src/compiler/interpreter.jl [hunk 1]
```diff
@@ -140,52 +137,121 @@ function isKWCallSignature(@nospecialize(TT))
...
+    elseif EnzymeRules.is_inactive_from_sig(specTypes; world = interp.world, method_table)
+        callinfo = NoInlineCallInfo(callinfo, atype, :inactive)
+    elseif interp.mode == API.DEM_ForwardMode
+        if EnzymeRules.has_frule_from_sig(specTypes; world = interp.world, method_table)
+            callinfo = NoInlineCallInfo(callinfo, atype, :frule)
+        end
+    elseif EnzymeRules.has_rrule_from_sig(specTypes; world = interp.world, method_table)
+        callinfo = NoInlineCallInfo(callinfo, atype, :rrule)
     end
+    return Core.Compiler.CallMeta(ret.rt, ret.exct, ret.effects, callinfo)
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:190]:
```suggestion
    @static if VERSION ≥ v"1.11-"
    return Core.Compiler.CallMeta(ret.rt, ret.exct, ret.effects, callinfo)
    else
    return Core.Compiler.CallMeta(ret.rt, ret.effects, callinfo)
    end
```

**Comment by wsmoses** [comment]
That feels unrelated (and likely on me to fix other misc 1.11 stuff)

**Comment by wsmoses** [comment]
This generally looks good to me, but seems to break 1.10 (which should pass atm — the lower versions have a Yggdrasil issue rn which shouldn’t block this landing)

**Comment by wsmoses** [comment]
@aviatesk I rebased this [now with all the 1.11 JLL issues remedied, and I must have done something improper as 1.10 fails.

Is it possible to get some help (and sorry about that)

File: src/compiler/interpreter.jl [hunk 2]
```diff
@@ -123,52 +120,81 @@ function isKWCallSignature(@nospecialize(TT))
...
+            @safe_debug "Blocking inlining due to rrule" info.tt
+        end
+        return nothing
+    elseif info isa AlwaysInlineCallInfo
+        @safe_debug "Forcing inlining for primitive func" info.tt
+        return src
+    end
+    return @invoke Core.Compiler.inlining_policy(interp::AbstractInterpreter,
+        src::Any, info::CallInfo, stmt_flag::UInt32)
 end
```
Comment (aviatesk) [inline src/compiler/interpreter.jl:198]:
```suggestion
let # overload `inlining_policy`
    @static if VERSION ≥ v"1.11.0-DEV.879"
        sigs_ex = :(analyzer::AbstractAnalyzer, @nospecialize(src), @nospecialize(info::Core.Compiler.CallInfo), stmt_flag::UInt32)
        args_ex = :(analyzer::AbstractInterpreter, src::Any, info::Core.Compiler.CallInfo, stmt_flag::UInt32)
    else
        sigs_ex = :(analyzer::AbstractAnalyzer,
            @nospecialize(src), @nospecialize(info::Core.Compiler.CallInfo), stmt_flag::UInt8, mi::MethodInstance, argtypes::Vector{Any})
        args_ex = :(analyzer::AbstractInterpreter,
            src::Any, info::Core.Compiler.CallInfo, stmt_flag::UInt8, mi::MethodInstance, argtypes::Vector{Any})
    end
    @eval function Core.Compiler.inlining_policy($(sigs_ex.args...))
        if info isa NoInlineCallInfo
            if info.kind === :primitive
                @safe_debug "Blocking inlining for primitive func" info.tt
            elseif info.kind === :inactive
                @safe_debug "Blocking inlining due to inactive rule" info.tt
            elseif info.kind === :frule
                @safe_debug "Blocking inlining due to frule" info.tt
            else
                @assert info.kind === :rrule
                @safe_debug "Blocking inlining due to rrule" info.tt
            end
            return nothing
        elseif info isa AlwaysInlineCallInfo
            @safe_debug "Forcing inlining for primitive func" info.tt
            return src
        end
        return @invoke Core.Compiler.inlining_policy($(args_ex.args...))
    end
end
```

**Comment by aviatesk** [comment]
Ok, I think the failure has been fixed.

**Pull request merged**